includes/class-wc-stripe-customer.php

```diff
@@ -242,30 +258,11 @@ public function add_source( $source_id ) {
 			$this->set_id( $this->create_customer() );
 		}
 
-		$response = WC_Stripe_API::request(
-			array(
-				'source' => $source_id,
-			),
-			'customers/' . $this->get_id() . '/sources'
-		);
+		$response = $this->attach_source( $source_id );
 
+		// Add token to WooCommerce.
 		$wc_token = false;
 
-		if ( ! empty( $response->error ) ) {
-			// It is possible the WC user once was linked to a customer on Stripe
-			// but no longer exists. Instead of failing, lets try to create a
-			// new customer.
-			if ( $this->is_no_such_customer_error( $response->error ) ) {
-				$this->recreate_customer();
-				return $this->add_source( $source_id );
-			} else {
-				return $response;
-			}
-		} elseif ( empty( $response->id ) ) {
-			return new WP_Error( 'error', __( 'Unable to add payment source.', 'woocommerce-gateway-stripe' ) );
-		}
-
-		// Add token to WooCommerce.
```

Comment:
If I'm reading the code right, I think we can't extract the entirety of this into `attach_source`, since `$response` could be in an erroneous state when returned from `attach_source`, in which case we should return `$response` immediately—like we did before this change.

So I think we need some sort of `if ( ! empty ( $response->error ) && ! $this->is_no_such_customer_error( $response->error ) ) { return $response; }` conditional here.

Alternatively we could also just not extract this into the `attach_source` function, but I do think it fits semantically into that function. I guess the question is what is the best way to handle this?

Reply:
Thanks for catching this, @reykjalin!

In my opinion the best approach here is to check whether `attach_source` returned an error and return that error from `add_source`. This preserves the original behavior of the method.

While testing this, I realized that the create-customer-if-not-existing can moved to `attach_source` from `add_source`, which eliminates the need to do it in `create_setup_intent` as well.

To test this:

On a customer with existing card:

1. Modify the `wp__stripe_customer_id` meta value in the `wp_usermeta` table in order to force an invalid customer.
2. Try adding a new card.

The meta value should be updated with the new customer, and the card should be added to that customer.